### ISSUE TITLE
Reverse desyncs when renshaturbo setting is enabled

### ISSUE THREAD

**Issue by MBilderbeek** [openMSX/openMSX#1003]
Hydragon reports that a replay he loaded appeared to be completely broken. Then he found out that in Catapult he had the RenshaTurbo setting enabled. After disabling it and reloading the replay, the replay worked fine.
In this case, the setting here is influencing the actual input the MSX receives! This was apparently not taken into account.

**Comment by Colpocorto** [comment]
I also found this error on openMSX 16.0-414.

The following video shows an example of an out of sync replay (as from minute 11:00 all the recorded input strokes are replayed at the wrong screens, sticking the game on the weapon menu instead of replaying the actual game):

https://youtu.be/7-qfj67dFk0

[mars.omr.zip](https://github.com/openMSX/openMSX/files/6291419/mars.omr.zip)

**Comment by MBilderbeek** [comment]
So, did you have renshaturbo enabled? Or is this a new unrelated issue with this bug report?
I can't load the OMR file, as I don't have the required ROM file:
`Fatal error: Cannot load replay: Couldn't find ROM file for "MSXRom" D:/MSX/TEMP/mars.rom (sha1: 65c72353a6e21f3140652d595957d32f9d357d26).`

Was there something special you did, like activating renshaturbo or something else that may explain the desync?

**Comment by Colpocorto** [comment]
> So, did you have renshaturbo enabled? Or is this a new unrelated issue with this bug report?
> I can't load the OMR file, as I don't have the required ROM file:
> `Fatal error: Cannot load replay: Couldn't find ROM file for "MSXRom" D:/MSX/TEMP/mars.rom (sha1: 65c72353a6e21f3140652d595957d32f9d357d26).`
> 
> Was there something special you did, like activating renshaturbo or something else that may explain the desync?

Yes, I activated the renshaturbo feature. The ROM is from Strategic Mars, by the way.
So, is this issue opened since 2016?

**Comment by MBilderbeek** [comment]
I understand the ROM is Strategic Mars, but I don't have that particular ROM image which is required for the replay.

I guess you already tried to replay it in view-only mode? (Just drop the omr file on openMSX.)

When did the problems start? When you enabled rensha-turbo? Or did you enable that right from the start?

Anyway, yes, this is an old issue. Better not use reverse in combination with renshaturbo. I'm not sure whether we can reasonably fix this, perhaps @m9710797 can comment.

**Comment by Colpocorto** [comment]
> I understand the ROM is Strategic Mars, but I don't have that particular ROM image which is required for the replay.
>
Oh yes... I modified the ROM to start in R800 mode. That's why the hash doesn't match.
 
> I guess you already tried to replay it in view-only mode? (Just drop the omr file on openMSX.)
>
Yes I tried that but no success.
 
> When did the problems start? When you enabled rensha-turbo? Or did you enable that right from the start?

I enabled the renshaturbo early during the play. However, the out of sync seemed to start about minute 10:00.
 
> Anyway, yes, this is an old issue. Better not use reverse in combination with renshaturbo. I'm not sure whether we can reasonably fix this, perhaps @m9710797 can comment.

Ok.

**Comment by m9710797** [comment]
Yes, it's an old known issue.
I'll put fixing it higher on my priority list. (But probably won't make it anymore for the 17.0 release).

I *assume* the problem is that changes in the 'renshaturbo' setting don't get stored in the replay file. If this assumption is correct that means that, after the bug is fixed, the old replay files still won't work in newer openMSX versions.

**Comment by mthuurne** [comment]
Should we be recording the renshaturbo setting though or the events that it generates? For an external autofire I think the events should be recorded. Is it different for internal or does it not really matter whether the circuit is internal or external?

**Comment by m9710797** [comment]
My idea was to record the changes in the renshaturbo-slider. So not directly the setting itself, but the effect that the setting has. Similar to how we not directly record keyboard events, but the effect that those have on the keyboard matrix.
In other words, sort of, make the slider part of the input status that must be recorded.

But I prefer to only do that after the 17.0 release (depending on how long that will still take).

**Comment by m9710797** [comment]
I pushed a fix. Now both the initial renshaturbo state and changes in this state should be properly recorded.
Unfortunately recordings made before this fix will still be wrong (will still desync when replayed with a different renshaturbo setting).

I did some basic testing, but could use some more testing. So I'm leaving this ticket still open.

**Comment by MBilderbeek** [comment]
I guess it's time to close it. Please reopen if it appears to be broken after all.
